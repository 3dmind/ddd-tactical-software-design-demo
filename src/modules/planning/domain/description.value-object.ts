import { Guard, Result } from '../../../shared/core';
import { ValueObject } from '../../../shared/domain';

interface DescriptionPropsInterface {
  value: string;
}

export class DescriptionValueObject extends ValueObject<
  DescriptionPropsInterface
> {
  public static minLength = 2;
  public static maxLength = 250;

  private constructor(props: DescriptionPropsInterface) {
    super(props);
  }

  get value(): string {
    return this.props.value;
  }

  public static create(text: string): Result<DescriptionValueObject> {
    const nullGuardResult = Guard.againstNullOrUndefined(text, 'text');

    if (!nullGuardResult.succeeded) {
      return Result.fail<DescriptionValueObject>(nullGuardResult.message);
    }

    const minGuardResult = Guard.againstAtLeast(this.minLength, text);
    const maxGuardResult = Guard.againstAtMost(this.maxLength, text);

    if (!minGuardResult.succeeded) {
      return Result.fail<DescriptionValueObject>(minGuardResult.message);
    }

    if (!maxGuardResult.succeeded) {
      return Result.fail<DescriptionValueObject>(maxGuardResult.message);
    }

    return Result.ok<DescriptionValueObject>(
      new DescriptionValueObject({ value: text }),
    );
  }
}
