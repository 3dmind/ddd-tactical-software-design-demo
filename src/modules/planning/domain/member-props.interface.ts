import { UserId } from '../../users/domain/user-id.entity';

export interface MemberProps {
  userId: UserId;
  createdAt?: Date;
}
