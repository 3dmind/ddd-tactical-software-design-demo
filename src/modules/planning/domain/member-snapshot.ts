import { UserId } from '../../users/domain/user-id.entity';
import { MemberId } from './member-id.entity';
import { MemberProps } from './member-props.interface';

export class MemberSnapshot {
  readonly createdAt: Date;
  readonly memberId: MemberId;
  readonly userId: UserId;

  constructor(props: Readonly<MemberProps>, memberId: MemberId) {
    const { createdAt, userId } = props;
    this.createdAt = createdAt;
    this.memberId = memberId;
    this.userId = userId;

    Object.freeze(this);
  }
}
