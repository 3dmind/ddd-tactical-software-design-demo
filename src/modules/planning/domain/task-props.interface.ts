import { Description } from './description.valueobject';

export interface TaskProps {
  archived: boolean;
  archivedAt: Date;
  createdAt: Date;
  description: Description;
  discarded: boolean;
  discardedAt: Date;
  editedAt: Date;
  resumedAt: Date;
  tickedOff: boolean;
  tickedOffAt: Date;
}
