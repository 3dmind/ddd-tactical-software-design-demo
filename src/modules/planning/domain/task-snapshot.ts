import { DescriptionValueObject } from './description.value-object';
import { TaskIdEntity } from './task-id.entity';
import { TaskPropsInterface } from './task-props.interface';

export class TaskSnapshot {
  readonly archivedAt: Date;
  readonly createdAt: Date;
  readonly description: DescriptionValueObject;
  readonly discardedAt: Date;
  readonly editedAt: Date;
  readonly isArchived: boolean;
  readonly isDiscarded: boolean;
  readonly isTickedOff: boolean;
  readonly resumedAt: Date;
  readonly taskId: TaskIdEntity;
  readonly tickedOffAt: Date;

  constructor(props: Readonly<TaskPropsInterface>, taskId: TaskIdEntity) {
    const {
      archived,
      archivedAt,
      createdAt,
      description,
      discarded,
      discardedAt,
      editedAt,
      resumedAt,
      tickedOff,
      tickedOffAt,
    } = props;
    this.archivedAt = archivedAt;
    this.createdAt = createdAt;
    this.description = description;
    this.discardedAt = discardedAt;
    this.editedAt = editedAt;
    this.isArchived = archived;
    this.isDiscarded = discarded;
    this.isTickedOff = tickedOff;
    this.resumedAt = resumedAt;
    this.taskId = taskId;
    this.tickedOffAt = tickedOffAt;

    Object.freeze(this);
  }
}
