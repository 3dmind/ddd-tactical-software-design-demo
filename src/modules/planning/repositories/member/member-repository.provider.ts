import { Provider } from '@nestjs/common';
import { MemberRepository } from './member.repository';
import { PrismaMemberRepository } from './prisma-member.repository';

export const MemberRepositoryProvider: Provider = {
  provide: MemberRepository,
  useClass: PrismaMemberRepository,
};
