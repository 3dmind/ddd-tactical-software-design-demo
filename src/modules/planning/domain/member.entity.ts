import { Guard, Result } from '../../../shared/core';
import { Entity, UniqueEntityId } from '../../../shared/domain';
import { UserId } from '../../users/domain/user-id.entity';
import { MemberId } from './member-id.entity';
import { MemberProps } from './member-props.interface';
import { MemberSnapshot } from './member-snapshot';

export class Member extends Entity<MemberProps> {
  private constructor(props: MemberProps, id?: UniqueEntityId) {
    super(props, id);
  }

  get memberId(): MemberId {
    return MemberId.create(this._id).getValue();
  }

  get userId(): UserId {
    return this.props.userId;
  }

  public static create(
    props: MemberProps,
    id?: UniqueEntityId,
  ): Result<Member> {
    const guardResult = Guard.againstNullOrUndefinedBulk([
      {
        argument: props.userId,
        argumentName: 'userId',
      },
    ]);

    if (!guardResult.succeeded) {
      return Result.fail<Member>(guardResult.message);
    }

    const member = new Member(
      {
        ...props,
        createdAt: props.createdAt ?? new Date(),
      },
      id,
    );
    return Result.ok<Member>(member);
  }

  public createSnapshot(): MemberSnapshot {
    return new MemberSnapshot(this.props, this.memberId);
  }
}
