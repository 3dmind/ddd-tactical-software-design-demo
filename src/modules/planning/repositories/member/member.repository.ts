import { UniqueEntityId } from '../../../../shared/domain';
import { Member } from '../../domain/member.entity';

export type MaybeMember = {
  found: boolean;
  member?: Member;
};

export abstract class MemberRepository {
  abstract async exists(id: UniqueEntityId): Promise<boolean>;

  abstract async getMemberByUserId(id: UniqueEntityId): Promise<MaybeMember>;

  abstract async save(member: Member): Promise<void>;
}
