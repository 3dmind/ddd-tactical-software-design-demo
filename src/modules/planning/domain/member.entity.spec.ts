import * as faker from 'faker';
import * as uuid from 'uuid';
import { UniqueEntityId } from '../../../shared/domain';
import { UserId } from '../../users/domain/user-id.entity';
import { UserName } from '../../users/domain/user-name.valueobject';
import { MemberProps } from './member-props.interface';
import { Member } from './member.entity';

jest.mock('uuid');

describe('Member entity', () => {
  beforeAll(() => {
    uuid.v4.mockReturnValue(faker.random.uuid());
  });

  afterAll(() => {
    uuid.mockRestore();
  });

  it('should guard "userId" property', () => {
    expect.assertions(4);
    const propsWithNull = { userId: null } as MemberProps;
    const propsWithUndefined = {} as MemberProps;

    const memberResultNull = Member.create(propsWithNull);
    const memberResultUndefined = Member.create(propsWithUndefined);

    expect(memberResultNull.isFailure).toBe(true);
    expect(memberResultNull.errorValue()).toEqual(
      'userId is null or undefined',
    );
    expect(memberResultUndefined.isFailure).toBe(true);
    expect(memberResultUndefined.errorValue()).toEqual(
      'userId is null or undefined',
    );
  });

  it('should create new member', () => {
    expect.assertions(2);
    const userIdFixture = UserId.create().getValue();
    UserName.create(faker.internet.userName()).getValue();
    const idFixture = faker.random.uuid();
    const entityIdFixture = new UniqueEntityId(idFixture);

    const memberResult = Member.create(
      {
        userId: userIdFixture,
      },
      entityIdFixture,
    );
    const member = memberResult.getValue();

    expect(memberResult.isSuccess).toBe(true);
    expect(member.memberId.id.toValue()).toEqual(idFixture);
  });

  it('should create snapshot', () => {
    expect.assertions(1);
    const userIdFixture = UserId.create(
      new UniqueEntityId('b21b86ba-a05c-4425-b8d7-09dda742199d'),
    ).getValue();
    UserName.create('TomTest').getValue();
    const dateFixture = new Date(Date.parse('1977-01-01'));
    const idFixture = new UniqueEntityId(
      '773f023e-3c57-4bd5-aa0a-0e0e0cc985ab',
    );
    const member = Member.create(
      {
        userId: userIdFixture,
        createdAt: dateFixture,
      },
      idFixture,
    ).getValue();

    const memberSnapshot = member.createSnapshot();

    expect(memberSnapshot).toMatchSnapshot();
  });
});
