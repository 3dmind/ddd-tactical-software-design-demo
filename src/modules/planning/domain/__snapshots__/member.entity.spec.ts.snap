// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Member entity should create snapshot 1`] = `
MemberSnapshot {
  "createdAt": 1977-01-01T00:00:00.000Z,
  "memberId": MemberId {
    "_id": UniqueEntityId {
      "value": "773f023e-3c57-4bd5-aa0a-0e0e0cc985ab",
    },
    "props": null,
  },
  "userId": UserId {
    "_id": UniqueEntityId {
      "value": "b21b86ba-a05c-4425-b8d7-09dda742199d",
    },
    "props": null,
  },
}
`;
